test/test_nn.py

```diff
@@ -533,7 +533,8 @@ def _backward_criterion(self, criterion, input, target, gradOutput=None, extra_a
     def _zero_grad_parameters(self, module):
         for p in module.parameters():
             if p.grad is not None:
-                p.grad.data.zero_()
+                with torch.no_grad():
+                    p.grad.zero_()
```

Comment:
From @gchanan:
This seems a bit fishy.  For example, couldn't you implement zero_() by not changing the shape of the metadata and have this working?  Obviously this would not be very efficient.

But I don't really understand what our guarantees are for SparseTensors with autograd -- @SsnL can you chime in?

Reply:
From @yf225 :

Behavior of some sparse tensor operations with/without `allow_tensor_metadata_change_` flag:

`sparse_tensor.zero_()`
```python
>>> import torch
>>> i = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v = torch.tensor([3., 4., 5.])
>>> st = torch.sparse.FloatTensor(i, v)
>>> dst = st.data

# Without the allow_tensor_metadata_change_ flag:
>>> dst.zero_()   # Note that `dst` and `st` have different values now
>>> dst
tensor(indices=tensor([], size=(2, 0)), values=tensor([], size=(0,)), size=(2, 3), nnz=0, layout=torch.sparse_coo)
>>> st
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([3., 4., 5.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)

# With the allow_tensor_metadata_change_ flag:
>>> dst.zero_()   # This operation is disallowed, because it breaks consistency between `dst` and `st`
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: resize_and_clear_ is not allowed on Tensor created from .data or .detach()
```

`sparse_tensor.copy_(st)`
```python
>>> import torch
>>> i = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v = torch.tensor([3., 4., 5.])
>>> st = torch.sparse.FloatTensor(i, v)
>>> dst = st.data

# Without the allow_tensor_metadata_change_ flag:
>>> i2 = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v2 = torch.tensor([10., 10., 10.])
>>> st2 = torch.sparse.FloatTensor(i2, v2)
>>> dst.copy_(st2)  # Note that `dst` and `st` have different values now
>>> dst
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([10., 10., 10.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)
>>> st
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([3., 4., 5.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)

# With the allow_tensor_metadata_change_ flag:
>>> dst.copy_(st2)   # This operation is disallowed, because it breaks consistency between `dst` and `st`
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: resize_ is not allowed on Tensor created from .data or .detach()
```

`sparse_tensor.add_(st)`
```python
>>> import torch
>>> i = torch.tensor([[0, 1, 1], [2, 0, 2]])
>>> v = torch.tensor([3., 4., 5.])
>>> st = torch.sparse.FloatTensor(i, v)
>>> dst = st.data
>>> st2 = torch.sparse.FloatTensor(torch.tensor([[0, 1, 1], [2, 0, 2]]), torch.tensor([10., 10., 10.]))

# Without the allow_tensor_metadata_change_ flag:
>>> dst.add_(st2)   # Note that `dst` and `st` have different values now
>>> dst
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([13., 14., 15.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)
>>> st
tensor(indices=tensor([[0, 1, 1], [2, 0, 2]]), values=tensor([3., 4., 5.]), size=(2, 3), nnz=3, layout=torch.sparse_coo)

# With the allow_tensor_metadata_change_ flag:
>>> dst.add_(st2)   # This operation is disallowed, because it breaks consistency between `dst` and `st`
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: set_indices_and_values_unsafe is not allowed on Tensor created from .data or .detach()
```

So the consistency-breaking changes can be prevented by the `allow_tensor_metadata_change_` flag.

Reply:
From @yf225 :

A better fix here might be the following, to reflect the different semantics for dense vs. sparse:
```python
if p.grad is not None:
    if p.is_sparse:  # `zero_()` on sparse tensor changes tensor metadata, and it should only be performed on the original tensor
        p.grad.zero_()
    else:
        p.grad.data.zero_()
```

Reply:
Conclusion from offline meeting:
1. We will use the fix in this PR.
2. We will open a Github issue to track this problem.

Reply:
Tracking issue for in-place op on sparse `.data` and `.detach()`: https://github.com/pytorch/pytorch/issues/15070